x-pack/plugins/ingest_manager/public/applications/ingest_manager/sections/agent_config/edit_datasource_page/index.tsx

```diff
@@ -0,0 +1,321 @@
+          // Remove `agent_stream` from all stream info, we assign this after saving
+          const newDatasource = {
+            ...restOfDatasource,
+            inputs: inputs.map(input => {
+              const { streams, ...restOfInput } = input;
+              return {
+                ...restOfInput,
+                streams: streams.map(stream => {
+                  const { agent_stream, ...restOfStream } = stream;
+                  return { ...restOfStream };
+                }),
+              };
+            }),
+          };
+          setDatasource(newDatasource);
+          if (datasourceData.item.package) {
+            const { data: packageData } = await sendGetPackageInfoByKey(
+              `${datasourceData.item.package.name}-${datasourceData.item.package.version}`
+            );
+            if (packageData?.response) {
+              setPackageInfo(packageData.response);
+              setValidationResults(validateDatasource(newDatasource, packageData.response));
+              setFormState('VALID');
+            }
+          }
+        }
+      } catch (e) {
+        setLoadingError(e);
+      }
+      setIsLoadingData(false);
+    };
+    getData();
+  }, [configId, datasourceId]);
+
+  // Retrieve agent count
+  const [agentCount, setAgentCount] = useState<number>(0);
+  useEffect(() => {
+    const getAgentCount = async () => {
+      const { data } = await sendGetAgentStatus({ configId });
+      if (data?.results.total) {
+        setAgentCount(data.results.total);
+      }
+    };
+
+    if (isFleetEnabled) {
+      getAgentCount();
+    }
+  }, [configId, isFleetEnabled]);
+
+  // Datasource validation state
+  const [validationResults, setValidationResults] = useState<DatasourceValidationResults>();
+  const hasErrors = validationResults ? validationHasErrors(validationResults) : false;
+
+  // Update datasource method
+  const updateDatasource = (updatedFields: Partial<NewDatasource>) => {
+    const newDatasource = {
+      ...datasource,
+      ...updatedFields,
+    };
+    setDatasource(newDatasource);
+
+    // eslint-disable-next-line no-console
+    console.debug('Datasource updated', newDatasource);
+    const newValidationResults = updateDatasourceValidation(newDatasource);
+    const hasValidationErrors = newValidationResults
+      ? validationHasErrors(newValidationResults)
+      : false;
+    if (!hasValidationErrors) {
+      setFormState('VALID');
+    }
+  };
+
+  const updateDatasourceValidation = (newDatasource?: NewDatasource) => {
+    if (packageInfo) {
+      const newValidationResult = validateDatasource(newDatasource || datasource, packageInfo);
+      setValidationResults(newValidationResult);
+      // eslint-disable-next-line no-console
+      console.debug('Datasource validation results', newValidationResult);
+
+      return newValidationResult;
+    }
+  };
+
+  // Cancel url
+  const CONFIG_URL = useLink(`${AGENT_CONFIG_DETAILS_PATH}${configId}`);
```

Comment:
I think we have a way to get this but either way, I would like to avoid views having knowledge of the route specifics and having to assemble them.